index.d.ts

```diff
@@ -0,0 +1,476 @@
+
+        /**
+         * Invoked to check if the requested scope is valid for a particular client/user combination.
+         *
+         */
+        validateScope?(user: User, client: Client, scope: string | string[], callback?: Callback<string | Falsey>): Promise<string | string[] | Falsey>;
+    }
+
+    interface PasswordModel extends BaseModel, RequestAuthenticationModel {
+        /**
+         * Invoked to generate a new refresh token.
+         *
+         */
+        generateRefreshToken?(client: Client, user: User, scope: string | string[], callback?: Callback<string>): Promise<string>;
+
+        /**
+         * Invoked to retrieve a user using a username/password combination.
+         *
+         */
+        getUser(username: string, password: string, callback?: Callback<User | Falsey>): Promise<User | Falsey>;
+
+        /**
+         * Invoked to check if the requested scope is valid for a particular client/user combination.
+         *
+         */
+        validateScope?(user: User, client: Client, scope: string | string[], callback?: Callback<string | Falsey>): Promise<string | string[] | Falsey>;
+    }
+
+    interface RefreshTokenModel extends BaseModel, RequestAuthenticationModel {
+        /**
+         * Invoked to generate a new refresh token.
+         *
+         */
+        generateRefreshToken?(client: Client, user: User, scope: string | string[], callback?: Callback<string>): Promise<string>;
+
+        /**
+         * Invoked to retrieve an existing refresh token previously saved through Model#saveToken().
+         *
+         */
+        getRefreshToken(refreshToken: string, callback?: Callback<RefreshToken>): Promise<RefreshToken | Falsey>;
+
+        /**
+         * Invoked to revoke a refresh token.
+         *
+         */
+        revokeToken(token: RefreshToken | Token, callback?: Callback<boolean>): Promise<boolean>;
+    }
+
+    interface ClientCredentialsModel extends BaseModel, RequestAuthenticationModel {
+        /**
+         * Invoked to retrieve the user associated with the specified client.
+         *
+         */
+        getUserFromClient(client: Client, callback?: Callback<User | Falsey>): Promise<User | Falsey>;
+
+        /**
+         * Invoked to check if the requested scope is valid for a particular client/user combination.
+         *
+         */
+        validateScope?(user: User, client: Client, scope: string | string[], callback?: Callback<string | Falsey>): Promise<string | string[] | Falsey>;
+    }
+
+    interface ExtensionModel extends BaseModel, RequestAuthenticationModel {}
+
+    /**
+     * An interface representing the user.
+     * A user object is completely transparent to oauth2-server and is simply used as input to model functions.
+     */
+    interface User {
+        [key: string]: any;
+    }
+
+    /**
+     * An interface representing the client and associated data
+     */
+    interface Client {
```

Comment:
Does the client not have a secret?

Reply:
Also should have user_id

Reply:
I can't find any reference to `client.secret` or `client.client_secret`. `user_id` only appears in spec.rst, but nowhere in the code.

Reply:
Actually I think the state is fine for now. We should actually merge it and fix stuff in an iterative manner.

Reply:
RFC 6749 https://datatracker.ietf.org/doc/html/rfc6749#section-2.3.1
client_id
client_secret

userId is part of oauth2_server specific implementation

Reply:
But where is it supported in this library?

Reply:
Let me think "loud":
You need it if you use confidential clients with ClientCredentialFlow. Then you use the secret in getClient. 

I think you are right: This is up to the implementer.